x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/profile/RestSuggestProfilesAction.java

```diff
@@ -55,16 +66,28 @@ public String getName() {
 
     @Override
     protected RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient client) throws IOException {
-        final Set<String> dataKeys = Strings.tokenizeByCommaToSet(request.param("data", null));
         final Payload payload = request.hasContentOrSourceParam()
             ? PARSER.parse(request.contentOrSourceParamParser(), null)
-            : new Payload(null, null);
+            : new Payload(null, null, null, null);
+
+        final String data = request.param("data", null);
+        if (data != null && payload.data() != null) {
+            throw new IllegalArgumentException(
+                "The [data] parameter must be specified in either request body or as query parameter, but not both"
```

Comment:
I think this should be covered already. The `payload` is parsed from `contentOrSourceParamParser` which covers both request body and `source` query parameter. As said above, if `source` is not consumed (because request body is provided), the whole request should error out.